java/cuvs-java/src/main/java22/com/nvidia/cuvs/internal/TieredIndexImpl.java

```diff
@@ -142,59 +119,54 @@ public void destroyIndex() throws Throwable {
    * @return an instance of {@link IndexReference} that holds the pointer to the
    *         index
    */
-  private IndexReference build() throws Throwable {
+  private IndexReference build() {
     try (var localArena = Arena.ofConfined()) {
-      long rows = dataset != null ? dataset.size() : vectors.length;
-      long cols = dataset != null ? dataset.dimensions() : (rows > 0 ? vectors[0].length : 0);
+      long rows = dataset.size();
+      long cols = dataset.dimensions();
 
       MemorySegment indexParamsMemorySegment =
           tieredIndexParameters != null
               ? segmentFromIndexParams(localArena, tieredIndexParameters)
               : MemorySegment.NULL;
 
       // Get host data
-      MemorySegment hostDataSeg =
-          dataset != null
-              ? ((DatasetImpl) dataset).asMemorySegment()
-              : Util.buildMemorySegment(localArena, vectors);
+      MemorySegment hostDataSeg = ((DatasetImpl) dataset).asMemorySegment();
 
-      long cuvsRes = resources.getHandle();
+      try (var resourceAccess = resources.access()) {
+        long cuvsRes = resourceAccess.handle();
 
-      // TieredIndex REQUIRES device memory - allocate it
-      MemorySegment datasetD = localArena.allocate(C_POINTER);
-      long datasetSize = C_FLOAT_BYTE_SIZE * rows * cols;
-      int returnValue = cuvsRMMAlloc(cuvsRes, datasetD, datasetSize);
-      checkCuVSError(returnValue, "cuvsRMMAlloc");
+        // TieredIndex REQUIRES device memory - allocate it
+        long datasetSize = C_FLOAT_BYTE_SIZE * rows * cols;
+        MemorySegment datasetDP = allocateRMMSegment(cuvsRes, datasetSize);
```

Comment:
Nitpick: Not to be handled in the current PR.

Do we stand the risk of leaking the allocation from RMM if there is an exception before the cleanup?  

Might need to look at this more closely, separately.